plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java

```diff
@@ -53,19 +51,26 @@
   /**
    * Returns the credential of an active user (among multiple logged-in users). A login screen
    * may be presented, e.g., if no user is logged in or login is required due to an expired
-   * credential.
+   * credential. This method can still return {@code null} if a user cancels the login process.
+   * Under normal circumstances, {@code null} is returned only when a user cancelled login.
+   * For this reason, if {@code null} is return, the caller is advised to cancel the current
+   * operation and display a general message that login is required but was cancelled or failed.
    *
    * @param shellProvider provides a shell for the login screen if login is necessary
-   * @throws IOException can be thrown by the underlying Login API request
+   * @throws IOException can be thrown by the underlying Login API request (e.g., network
+   *     error from the transport layer while sending/receiving a HTTP request/response.)
    */
-  // TODO(chanseok); synchronize properly
   public Credential getActiveCredential(IShellProvider shellProvider) throws IOException {
-    MApplication app = PlatformUI.getWorkbench().getService(MApplication.class);
-    Credential credential = (Credential) app.getTransientData().get(STASH_OAUTH_CRED_KEY);
+    MApplication application = PlatformUI.getWorkbench().getService(MApplication.class);
+
+    // get() and put() are not transactional, but we do our best to improve thread-safety.
```

Comment:
I agree, and I don't worry about a get followed by a put not locked together here. It's tolerable. (I saw FindBugs generates a warning in this case not being an atomic transaction.) Just wanted to make a note, as someone else who reads the code later might otherwise wonder why the original code writer didn't consider it while making the map thread-safe.